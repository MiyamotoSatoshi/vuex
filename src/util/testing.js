import Vue from 'vue'
import load from '~util/load'
import { mount, shallow } from 'avoriaz'

export function test(name, cb) {
  Vue.prototype.$vuetify = { load }

  beforeEach(() => {
    rafPolyfill(window)
  })

  describe(name, () => cb({
    functionalContext,
    mount,
    shallow
  }))
}

export function functionalContext(context = {}) {
  return {
    context: Object.assign({
      data: {},
      props: {}
    }, context)
  }
}


//requestAnimationFrame polyfill | Milos Djakonovic ( @Miloshio ) | MIT | https://github.com/milosdjakonovic/requestAnimationFrame-polyfill
export function rafPolyfill(w) {
  /**
   *
   * How many times should polyfill call
   * update callback? By canon, it should
   * be 60 times per second, so that ideal
   * framerate 60fps could be reached.
   *
   * However, even native implementations
   * of requestAnimationFrame often cannot
   * do 60fps, but, unlike any polyfill,
   * they can synchronise achievable fps
   * rate with screen refresh rate.
   *
   * So, leave this value 1000/60 unless
   * you target specific browser on spec
   * ific device that is going to work
   * better with custom value. I think
   * that this is the longest comment I've
   * written on single variable so far.
  **/
  var FRAME_RATE_INTERVAL = 1000/60,

  /**
   * All queued callbacks in given cycle.
  **/
  allCallbacks = [],

  executeAllScheduled = false,

  shouldCheckCancelRaf = false,

  /**
   * Callbacks queued for cancellation.
  **/
  callbacksForCancellation = [],

  /**
   * Should callback be cancelled?
   * @param cb - callback
  **/
  isToBeCancelled = function(cb){
    for(var i=0;i<callbacksForCancellation.length;i++){
      if(callbacksForCancellation[i] === cb ){
        callbacksForCancellation.splice(i,1);
        return true;
      }
    }
  },



  /**
   *
   * Executes all (surprise) callbacks in
   * and removes them from callback queue.
   *
  **/
  executeAll = function(){
    executeAllScheduled = false;
    var _allCallbacks = allCallbacks;
    allCallbacks = [];
    for(var i=0;i<_allCallbacks.length;i++){
      if(shouldCheckCancelRaf===true){
        if (isToBeCancelled(_allCallbacks[i])){
          shouldCheckCancelRaf = false;
          return;
        }
      }
      _allCallbacks[i].apply(w, [ new Date().getTime() ] );
    }
  },

  /**
   *
   * requestAnimationFrame polyfill
   * @param callback - callback to be queued & executed | executed
   * @return callback
   *
  **/
  raf = function(callback){
    allCallbacks.push(callback);
    if(executeAllScheduled===false){
      w.setTimeout(executeAll, FRAME_RATE_INTERVAL);
      executeAllScheduled = true;
    }
    return callback;
  },

  /**
   *
   * Cancels raf.
  **/
  cancelRaf = function(callback){
    callbacksForCancellation.push(callback);
    shouldCheckCancelRaf = true;
  },


  //https://gist.github.com/paulirish/1579671
  vendors = ['ms', 'moz', 'webkit', 'o'];

    for(var x = 0; x < vendors.length && !w.requestAnimationFrame; ++x) {
        w.requestAnimationFrame = w[vendors[x]+'RequestAnimationFrame'];
        w.cancelAnimationFrame = w[vendors[x]+'CancelAnimationFrame']
        || w[vendors[x]+'CancelRequestAnimationFrame'];
    }

  if (!w.requestAnimationFrame) w.requestAnimationFrame = raf;
  if (!w.cancelAnimationFrame)  w.cancelAnimationFrame  = cancelRaf;

}
