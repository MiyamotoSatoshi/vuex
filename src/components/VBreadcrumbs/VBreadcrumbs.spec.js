import { test } from '~util/testing'
import {
  VBreadcrumbs,
  VBreadcrumbsItem
} from '~components/VBreadcrumbs'

test('VBreadcrumbs.js', ({ mount }) => {
  it('should have breadcrumbs classes', () => {
    const wrapper = mount(VBreadcrumbs, {
      propsData: { icons: true }
   })

    expect(wrapper.hasClass('breadcrumbs')).toBe(true)
    expect(wrapper.hasClass('breadcrumbs--with-icons')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should inject divider to children', () => {
    const wrapper = mount(VBreadcrumbs, {
      slots: {
        default: [VBreadcrumbsItem]
      }
    })

    const item = wrapper.find(VBreadcrumbsItem)[0]
    expect(item.getAttribute('data-divider')).toBe('/')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
