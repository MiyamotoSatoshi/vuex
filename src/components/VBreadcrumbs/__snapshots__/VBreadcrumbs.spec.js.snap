// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VBreadcrumbs.js should have breadcrumbs classes 1`] = `

<ul class="breadcrumbs">
</ul>

`;

exports[`VBreadcrumbs.js should inject slot to children 1`] = `

<ul class="breadcrumbs">
  <li>
    <a class="breadcrumbs__item">
    </a>
  </li>
  <li class="breadcrumbs__divider">
    /
  </li>
  <li>
    <a class="breadcrumbs__item">
    </a>
  </li>
  <li class="breadcrumbs__divider">
    /
  </li>
  <li>
    <a class="breadcrumbs__item">
    </a>
  </li>
  <li class="breadcrumbs__divider">
    /
  </li>
  <li>
    <a class="breadcrumbs__item">
    </a>
  </li>
</ul>

`;

exports[`VBreadcrumbs.js should use a custom divider slot 1`] = `

<ul class="breadcrumbs">
  <li>
    <a class="breadcrumbs__item">
    </a>
  </li>
  <li class="breadcrumbs__divider">
    /divider/
  </li>
  <li>
    <a class="breadcrumbs__item">
    </a>
  </li>
</ul>

`;

exports[`VBreadcrumbs.js should use custom justify props 1`] = `

<ul class="breadcrumbs">
</ul>

`;

exports[`VBreadcrumbs.js should use custom justify props 2`] = `

<ul class="breadcrumbs">
</ul>

`;
